docs/design/git-submodule-storage.md

```diff
@@ -0,0 +1,138 @@
+
+Git's protocol is designed for communicating between copies of the same
+repository. Notably, a Git fetch calculates the list of required objects by
+performing reachability checks between the refs on the local and the remote
+side. We should expect that this will only work well if the submodule repository
+is stored as a local Git repository.
+
+Rolling our own Git fetch is too complex to be worth the effort.
+
+### "jj op restore" and operation log format
+
+We want `jj op restore` in the superproject to restore to an "expected" state in
+the submodule, though the definition of "expected" is up for debate. At the very
+least, `jj op restore` should restore the submodule working copy, but we may
+want to support the full list of operations at some point.
+
+Currently, the operation log only references objects and refs in the
+superproject, so it is likely that proposed approaches will need to extend this
+format. It is also worth considering that submodules may be added, updated or
+removed in superproject commits, thus the list of submodules is likely to change
+over the repository's lifetime.
+
+### Nested submodules
+
+Git submodules may contain submodules themselves, so our chosen storage schemes
+should support that.
+
+### Bonus: supporting future extensions
+
+There are certain extension we may want to make in the future, but are non-goals
+today. When evaluating approaches, consider these as low-priority "bonus"
+points; they shouldn't weight heavily in our decision-making, but all else being
+equal, we should prefer the approach that is more easily extended.
+
+These extensions are:
+
+- Non-git subrepos
+- Colocated Git repos
+
+## Possible approaches
+
+### Approach 1: Git repos in the main Git backend
+
+Since the Git backend contains a Git repository, an 'obvious' default would be
+to store them in the Git superproject the same way Git does, i.e. in
+`.git/modules`. This should allow us to _mostly_ reuse the existing backend
+interfaces. However, it is not likely that all backends will support Git
+submodules (see "Approaches Rejected"), and we will have to adjust the backend
+interfaces accordingly.
+
+TODO(chooglen): Discuss operation log
+TODO(chooglen): Discuss nested submodules
+
+TODO(chooglen) I'm not sure the backend interfaces mean that the rest of jj can
+  just 'ignore' the existence of Git submodules, e.g. when populating the
+  working copy, can we just ask the Git backend to give the tree structure we
+  need without knowing about submodules?
+
+### Approach 2: Store Git submodules as full jj repos
+
+This would be somewhere in `.jj` but outside of `.jj/store`. We would then
+expose a "submodules" interface that gets hooked up to the relevant machinery
+(e.g. updating the working copy).
+
+TODO(chooglen): Discuss operation log
+TODO(chooglen): Discuss nested submodules
+
+### Approach 3: Store Git submodules as alternate Git backends
+
+First, teach jj to use multiple backends in a way that works for Git submodules.
```

Comment:
FWIW, I don't understand what this means. Which backend(s) do you mean? The commit backend only? Or also commit index, operation log, and working copy?

Reply:
> The commit backend only?

I meant this, and...

> Or also commit index, operation log,

I wasn't sure if that also implied that we'd also need additional commit indexes and operation logs. I'd imagine we would need them.

> working copy

I hadn't considered this at all. I don't know if we need one working copy per submodule.

Reply:
> I don't know if we need one working copy per submodule.

I don't think we do :) But in this solution, the working copy is owned by all the submodules (and the superproject) together, IIUC, so it seems that the working-copy implementations will need to know to not overwrite each other. But again, I don't really understand this proposal anyway, so it's quite likely that I just misunderstood.

Reply:
> But in this solution, the working copy is owned by all the submodules (and the superproject) together, IIUC, so it seems that the working-copy implementations will need to know to not overwrite each other.

Yes, I think that's an accurate summary. When populating the working copy from the superproject, we have to notice when a subdirectory is 'owned' by a submodule and leave it unpopulated (maybe except for adding a `.git` file). Then, we need to go through the list of subdirectories and figure out which commit to use and from which backend.